### ISSUE TITLE
Creating Relationships works, but they are empty can I cannot recalculate

### ISSUE THREAD

**Issue by edhans** [TabularEditor/TabularEditor3#957]
**Describe the issue**
I can now create relationships in TE/3 for Power BI Desktop, but the visuals return the error that it cannot render the visual because a relationship is empty and needs to be recalculated.

I cannot recalculate the model when it is PBID, and unlike creating/editing calculated tables/columns, PBID doesn't tell you that you need to recalculate data and give you a button to do so. 

Refreshing visuals does nothing. Other than a full data refresh, I cannot figure out how to calculate a relationship without the workaround at the bottom of this post.

Not sure if this is a TE/3 bug, or a shortcoming in PBID. 

** Which version are you currently using? **
Tabular Editor 3.9, Enterprise

**To Reproduce**
Power BI Desktop

1. Create a relationship using Tabular Editor and save it to the model
2. Use a field from the DIM table (calendar in this example) and a measure from the fact table (Sales in this example) and create a simple Table viz
3. You get a gray box. See image below

**Expected behavior**
I would expect to get a dialog box like this to show up after saving the changes back to the model
<img width="469" alt="image" src="https://github.com/TabularEditor/TabularEditor3/assets/45298358/4b47319e-c474-46c3-92d3-47df1d4d98c6">

**Screenshots**
<img width="358" alt="image" src="https://github.com/TabularEditor/TabularEditor3/assets/45298358/22c9f632-5428-4168-9128-f9630ab5db89">

**Desktop (please complete the following information):**
Windows 11

**Additional context**
When I change a Calculated Table/Column and get the "refresh" yellow bar, pressing Refresh Now also refreshes the relationship, so it does work, but saving the relationship itself isn't triggering that dialog to come up.

**Comment by DBojsen** [comment]
Hi @christianwade

What do you think - shouldn't Power BI Desktop detect that a recalculation is needed and prompt the user, when adding a relationship from an external tool?
It would make sense to me and be concistent with adding calculated table as @edhans documented above.

BR David

**Comment by christianwade** [comment]
Makes sense. Like detect when need to do a process recalc for calculation groups. I'll forward the suggestion anyway ...

**Comment by dpesante** [comment]
I am running into this issue. Has this been corrected or can we force it as the last line in the c# script. I switch to using XMLA End Point and still seeing the same issue.

Any updates on this issue?

**Comment by otykier** [comment]
@dpesante if you're connected to the XMLA endpoint, you should be able to perform a Calculate Refresh through TE3, no?

Which C# script are you referring to?

**Comment by dpesante** [comment]
@otykier we have a c# script that we have been using that updates the Model (add calculated columns, create calc groupd, measures, etc.)... 

Currently, this is the last line I have in my c# script. Should I add another statement after that to force the refresh?

Model.AllMeasures.FormatDax();

**Comment by otykier** [comment]
I see, thanks for clarifying.

Yes, you can force a refresh by adding the following to your script:

```csharp
var type = "calculate";
var database = Model.Database.Name;
var tmsl = "{ \"refresh\": { \"type\": \"%type%\", \"objects\": [ { \"database\": \"%db%\" } ] } }"
    .Replace("%type%", type)
    .Replace("%db%", database);

ExecuteCommand(tmsl);
```

(**Note: This won't work against Power BI Desktop**)

See the [documentation](https://docs.tabulareditor.com/te2/Useful-script-snippets.html#querying-analysis-services) to learn more about the `ExecuteCommand` method.

**Comment by dpesante** [comment]
Great... that worked!!
